File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
from collections import deque
N = [0]
f1 = lambda v: v + N[0]
f2 = lambda v: v * 2
f3 = lambda v: v * 3
ls = [f1, f2, f3]

def bfs(x, y, n):
    que = deque()
    que.append((x, 0))
    visited = set()
    visited.add(x)
    
    while que:
        nx, cnt = que.popleft()
        if nx == y:
            return cnt
        
        for f in ls:
            ny = f(nx)
            if ny <= y and ny not in visited:
                visited.add(ny)
                que.append((ny, cnt + 1))
    
    return -1
    
    
def solution(x, y, n):
    N[0] = n
    return bfs(x,y,n)
    
        
        

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int x, int y, int n) {
        return bfs(x, y, n);
    }
    
    int bfs(int x, int y, int n){
        LinkedList<int[]> list = new LinkedList();
        Set<Integer> visited = new HashSet();
        visited.add(x);
        list.addLast(new int[]{x, 0});
        
        while(!list.isEmpty()){
            int[] nv = list.pollFirst();
            int nx = nv[0];
            int cnt = nv[1];
            
            if(nx == y) return cnt;
            
            for(int ny :getNextY(nx, n)){
                if (ny <= y && !visited.contains(ny)){
                    visited.add(ny);
                    list.addLast(new int[]{ny, cnt + 1});
                }
            }
        }
        
        return -1;
    }
    
    int[] getNextY(int x, int n){
        return new int[]{x + n, x * 2, x * 3};
    }
}